server/s3_put_multiobject_copy_action.cc

```diff
@@ -0,0 +1,962 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * For any questions about this software or licensing,
+ * please email opensource@seagate.com or cortx-questions@seagate.com.
+ *
+ */
+
+#include <cassert>
+#include <algorithm>
+#include <utility>
+
+#include "s3_common.h"
+#include "s3_common_utilities.h"
+#include "s3_put_multiobject_copy_action.h"
+#include "s3_error_codes.h"
+#include "s3_log.h"
+#include "s3_option.h"
+#include "s3_perf_logger.h"
+#include "s3_perf_metrics.h"
+#include "s3_stats.h"
+#include "s3_motr_layout.h"
+#include "s3_m0_uint128_helper.h"
+#include "s3_probable_delete_record.h"
+#include "s3_uri_to_motr_oid.h"
+
+extern struct s3_motr_idx_layout global_probable_dead_object_list_index_layout;
+
+const uint64_t MaxCopyObjectSourceSize = 5368709120UL;  // 5GB
+
+const char InvalidRequestSourceObjectSizeGreaterThan5GB[] =
+    "The specified copy source is larger than the maximum allowable size for a "
+    "copy source: 5368709120";
+const char InvalidRequestSourceAndDestinationSame[] =
+    "This copy request is illegal because it is trying to copy an object to "
+    "itself without changing the object's metadata, storage class, website "
+    "redirect location or encryption attributes.";
+
+const char DirectiveValueReplace[] = "REPLACE";
+const char DirectiveValueCOPY[] = "COPY";
+
+S3PutMultipartCopyAction::S3PutMultipartCopyAction(
+    std::shared_ptr<S3RequestObject> req, std::shared_ptr<MotrAPI> motr_api,
+    std::shared_ptr<S3ObjectMultipartMetadataFactory> object_mp_meta_factory,
+    std::shared_ptr<S3PartMetadataFactory> part_meta_factory,
+    std::shared_ptr<S3BucketMetadataFactory> bucket_meta_factory,
+    std::shared_ptr<S3ObjectMetadataFactory> object_meta_factory,
+    std::shared_ptr<S3MotrWriterFactory> motr_s3_writer_factory,
+    std::shared_ptr<S3MotrReaderFactory> motr_s3_reader_factory,
+    std::shared_ptr<S3MotrKVSWriterFactory> kv_writer_factory,
+    std::shared_ptr<S3AuthClientFactory> auth_factory)
+    : S3PutObjectActionBase(std::move(req), std::move(bucket_meta_factory),
+                            std::move(object_meta_factory), std::move(motr_api),
+                            std::move(motr_s3_writer_factory),
+                            std::move(kv_writer_factory)){
+  part_number = get_part_number();
+  upload_id = request->get_query_string_value("uploadId");
+  s3_log(S3_LOG_DEBUG, request_id, "%s Ctor\n", __func__);
+  s3_log(S3_LOG_INFO, stripped_request_id,
+         "S3 API: UploadPartCopy. Destination: [%s], Source: [%s]\
```

Comment:
Bucket name will be there. It will come like this:
`S3 API: UploadPartCopy. Destination: [testbkt3/someobject], Source: [testbkt1/object1] Part[1] for UploadId [ae630720-1e02-420f-98cf-885d9ffb4e4d]` 
Do you want to separate out bucket name?

Reply:
ok thats fine, as far as src & destination buckets can be identified, also looks compact